### PR TITLE
[BUGFIX beta] [PERF] Cache FactoryManagers

### LINKED ISSUE #15007: cache FactoryManager instances
[FactoryManager](https://github.com/emberjs/ember.js/blob/ef8f46e15d920dfbe008594a0177a9a2103f90ed/packages/container/lib/container.js#L663) should likely be cached, as it seems strange to create new ones for each instance.

In addition, we could cache the result of `makeToString` on it, so we don't need to create so many of those either.

cc @chadhietala

### PR DESCRIPTION
**Pull request by stefanpenner**
* 1 FactoryManager per container/normalizeFullName
* cache madeToString per factoryManager

[fixes #15007]

- [x] add tests – @chadhietala can you do this?

### CODE REVIEW

File: packages/container/lib/container.js [hunk 1]
```diff
@@ -289,16 +291,23 @@ if (isFeatureEnabled('ember-factory-for')) {
...
     }
 
+    let cached = this.factoryManagerCache[normalizedName];
+
+    if (cached) { return cached; }
+
     let factory = this.registry.resolve(normalizedName);
 
-    if (factory === undefined) { return; }
+    if (factory === undefined) {
```
Comment (krisselden) [inline packages/container/lib/container.js]: should misses not cache?

Comment (stefanpenner) [inline packages/container/lib/container.js]: i thought so, but there where test failures and I'm tired :P Maybe i can convince  @chadhietala to investigate?

Comment (stefanpenner) [inline packages/container/lib/container.js]: `resolve` does also cache so its not the end of the world either..

**Review by chadhietala — APPROVED** [review]

**Comment by chadhietala** [comment]
Added tests to make sure the `FactoryManager` instances are the same on a cache hit and are not equal on a cache miss.

File: packages/container/lib/container.js [hunk 2]
```diff
@@ -38,6 +38,7 @@ export default function Container(registry, options) {
   this.owner           = options && options.owner ? options.owner : null;
   this.cache           = dictionary(options && options.cache ? options.cache : null);
   this.factoryCache    = dictionary(options && options.factoryCache ? options.factoryCache : null);
+  this.factoryManagerCache = dictionary(options && options.factoryCache ? options.factoryCache : null);
```
Comment (rwjblue) [inline packages/container/lib/container.js]: This should not share the same cache as `this.factoryCache`.  I suspect this should be:

```js
this.factoryManagerCache = dictionary(options && options.factoryManagerCache ? options.factoryManagerCache : null);
```

Comment (rwjblue) [inline packages/container/lib/container.js]: Pushed a commit fixing this

**Comment by stefanpenner** [comment]
Thanks @chadhietala

**Comment by stefanpenner** [comment]
benchmarks (4.7x improvement):

```
node factory/factory-for.js                                                                                                 (1) (432ms)
testing
- factoryFor (new)
- factoryFor (old)
running first test, please wait...
  factoryFor (new) .... 4,342,542.07 op/s
  factoryFor (old) ...... 915,862.86 op/s
```

**Pull request merged**